### PR TITLE
perf: concurrent validation (TG-43)

### PR DESCRIPTION
**Pull request by Alyks1**
Signed-off-by: Alexander Gemmingen <alexander.gemmingen@gmx.de>

### CODE REVIEW

File: champ/node/validation/block.rs [hunk 1]
```diff
@@ -99,36 +100,41 @@ async fn verify_transactions(
     state: &ChampStateArc,
 ) -> Result<(), BlockValidationError> {
     debug!("verify transactions");
-    let db = &state.db.lock().await;
     // go through all tx in the block and do math to see new balance
     // check against block balance
-    let new_data = new_block.data.as_ref().ok_or(Node::BlockDataNotFound)?;
```
Comment (explodingcamera) [inline champ/node/validation/block.rs:105]: 🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀 finally as_ref is gone 🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀

File: champ/node/validation/block.rs [hunk 2]
```diff
@@ -99,36 +100,41 @@ async fn verify_transactions(
...
-        let tx_type = transaction.data.as_ref().ok_or(Validation::TransactionDataNotFound)?;
-        new_balance += match tx_type {
-            Data::TxSend(t) => -(t.amount as i128),
-            Data::TxCollect(t) => validate_collect(t, db, new_block).await?,
-            _ => new_balance,
-        };
+        let s = state.clone();
+        let trx = t_action.clone();
+        let block = new_block.clone();
+        let balance = new_balance.clone();
```
Comment (explodingcamera) [inline champ/node/validation/block.rs]:
```suggestion
        let balance = new_balance;
```

File: champ/node/validation/block.rs [hunk 3]
```diff
@@ -99,36 +100,41 @@ async fn verify_transactions(
...
     if new_data.transactions.len() > 255 {
         return Err(Validation::TooManyTransactions.into());
     }
 
-    // TODO: Run concurrently
     let mut new_balance: i128 = prev_data.balance as i128;
-    for transaction in &new_data.transactions {
+    let mut tokio_tasks: Vec<JoinHandle<Result<i128, BlockValidationError>>> = vec![];
+
+    for t_action in &new_data.transactions {
```
Comment (explodingcamera) [inline champ/node/validation/block.rs]: maybe change this to tx?

File: champ/node/validation/block.rs [hunk 4]
```diff
@@ -99,36 +100,41 @@ async fn verify_transactions(
...
+        let block = new_block.clone();
+        let balance = new_balance.clone();
+        // concurrent verification
+        let task: JoinHandle<Result<i128, BlockValidationError>> =
+            tokio::spawn(async move { trx_verification(&s, block, &balance, &trx).await });
+        tokio_tasks.push(task);
+    }
+
+    for t in tokio_tasks {
+        new_balance += t.await.map_err(|_| Node::AsyncError)??;
```
Comment (explodingcamera) [inline champ/node/validation/block.rs:137]: I like the ?? ?? ?? ?? ?? Really adds a nice ring to it.

**Review by explodingcamera — CHANGES_REQUESTED** [review]

File: champ/node/validation/block.rs (hunk 3 above)
Comment (Alyks1) [reply champ/node/validation/block.rs]: I use tx later so transactions makes more sense.

**Review by explodingcamera — APPROVED** [review]

**Pull request merged**